docs/design/AnoncredsW3CCompatibility.md

```diff
@@ -0,0 +1,286 @@
+        ),
+    )
+...
+```
+
+- W3C Credential Request
+
+```
+async def create_w3c_credential_request(
+        self, credential_offer: dict, credential_definition: CredDef, holder_did: str
+    ) -> Tuple[str, str]:
+...
+...
+try:
+  secret = await self.get_master_secret()
+  (
+      cred_req,
+      cred_req_metadata,
+  ) = await asyncio.get_event_loop().run_in_executor(
+      None,
+      W3CCredentialRequest.create,
+      None,
+      holder_did,
+      credential_definition.to_native(),
+      secret,
+      AnonCredsHolder.MASTER_SECRET_ID,
+      credential_offer,
+  )
+...
+```
+
+- W3C Credential Present
+
+```
+async def create_w3c_presentation(
+        self,
+        presentation_request: dict,
+        requested_credentials: dict,
+        schemas: Dict[str, AnonCredsSchema],
+        credential_definitions: Dict[str, CredDef],
+        rev_states: dict = None,
+    ) -> str:
+...
+...
+  try:
+    secret = await self.get_master_secret()
+    presentation = await asyncio.get_event_loop().run_in_executor(
+        None,
+        Presentation.create,
+        presentation_request,
+        present_creds,
+        self_attest,
+        secret,
+        {
+            schema_id: schema.to_native()
+            for schema_id, schema in schemas.items()
+        },
+        {
+            cred_def_id: cred_def.to_native()
+            for cred_def_id, cred_def in credential_definitions.items()
+        },
+    )
+...
+```
+
+#### Converting an already issued legacy anoncreds to W3C format(vice versa)
+
+In this case, we can use `to_w3c` method of `Credential` class to convert from legacy to w3c and `to_legacy` method of `W3CCredential` class to convert from w3c to legacy.<br>
+
+We could call `to_w3c` method like this:
+
+```
+w3c_cred = Credential.to_w3c(cred_def)
+```
+
+and for `to_legacy`:
+
+```
+legacy_cred = W3CCredential.to_legacy()
+```
+
+We don't need to input any parameters to it as it in turn calls `Credential.from_w3c()` method under the hood
+
+### How a W3C credential is stored in the wallet.
+
+Storing a credential in the wallet is somewhat dependent on the kinds of metadata that are relevent. The metadata mapping between the W3C credential and an anoncreds credential is not fully clear yet.
+
+One of the questions we need to answer is whether the preferred approach is to modify the existing store credential function so that any credential type is a valid input, or whether there should be a special function just for storing W3C credentials.
+
+We will duplicate this [store_credential](https://github.com/hyperledger/aries-cloudagent-python/blob/8cfe8283ddb2a85e090ea1b8a916df2d78298ec0/aries_cloudagent/anoncreds/holder.py#L167) function and modify it:
+
+```
+async def store_w3c_credential(...) {
+    ...
+    ...
+    try:
+        cred = W3CCredential.load(credential_data)
+    ...
+    ...
+}
+```
+
+**Question: Would it also be possible to generate the credentials on the fly to eliminate the need for storage?**
```

Comment:
If I understand this correctly, I don't think this is a viable approach -- particularly for a holder. In storing a credential, a big need is the ability to query for it.

Reply:
Querying is done against metadata, I think, not by peering inside actual credential content. So it might be possible to generate and index metadata without storing the content. But perhaps this is a distinction without a real difference, and maybe it's not worth entertaining. Storing both formats of a credential isn't hard.

One wrinkle is that, if a credential is indexed in two formats with essentially the same metadata, many wallets may suddenly find themselves having to prompt the user to select a format, when before they only got one hit from a query. If you know that the only credential that satisfies a query is your driver's license, and you get used to that credential being selected automatically, it may be a little jarring to suddenly have to decide which format of your credential you want to use. So I think one important to-do item might be to detect which format the proof request wants, and use that to filter. Proof requests that say, "I'll take either format" need to declare a preference for one over the other, or else agents need to have one format that they prefer over the other by default.

Reply:
Aries Agents are designed to store the credential as a blob, and to add tags to enable querying.  I did some checking with @andrewwhitehead about how that worked in ACA-Py/Askar/AnonCreds, and he let me know that it is ACA-Py's responsibility to extract/define the tags that point to an object (e.g., the credential), and to pass them to Askar for storage, along with the object itself. So, extracting the code, and adding a tag for the format(s) of a credential should be doable. No need to have two sets of tags for the same object just because it can be used in different types of presentations.